### DISCUSSION TITLE
What are the conditions for a service graph virtual node to exist?

### DISCUSSION THREAD

**Discussion by ogxd** [grafana/tempo#3480] [Q&A] [answered]
Hello!

In the [Tempo Service Graph documentation](https://grafana.com/docs/tempo/latest/metrics-generator/service_graphs/#virtual-nodes), it is said that 

> Virtual nodes can be detected [...] A client span does not have its matching server span, but has a peer attribute present [...] The default peer attributes are `peer.service`, `db.name` and `db.system`.

I do have many client spans with `peer.service` set, however I can't see any corresponding node in the service graph.
<img width="1089" alt="image" src="https://github.com/grafana/tempo/assets/12692438/67404893-8b64-4f6e-b1cb-178e444f8d30">

I do however have nodes for SQL server for which I have enabled automatic instrumentation, which sets `peer.service`, `db.system` and `db.name`. I notice that the node name corresponds to the `db.name` here.
![image](https://github.com/grafana/tempo/assets/12692438/0fa1bb6b-35c1-4326-a707-1bbbde6dcbf3)

Do you know why I have an "AdDiffusionRun" node but no "TensorFlowServing" node in my service graph? It's not a database so I don't want to set `db.name`.

**Comment by ogxd** [comment, marked as answer]
Ah I think I found the answer myself: I am using the opentelemetry-contrib servicegraphconnector and [the attributes are differen](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/connector/servicegraphconnector#configuration)t than the ones on the tempo documentation. 

> `virtual_node_peer_attributes`: the list of attributes need to match for building virtual server node, the higher the front, the higher the priority.
Default: [`db.name`, `net.sock.peer.addr`, `net.peer.name`, `rpc.service`, `net.sock.peer.name`, `net.peer.name`, `http.url`, `http.target`]

**Reply by ogxd** [reply]
Not sure if worth an issue. It can be misleading but this connector is a specific component and using it requires to go through the documentation anyway

**Reply by joe-elliott** [reply]
The metrics generation in the otel collector and Tempo have slowly drifted apart which is unfortunate. We would like them to be as in sync as possible but it's difficult to coordinate the two codebases.

Are you thinking the default virtual node peer attributes in Tempo or the collector make more sense? The otel collector likely has our original list which we found generated too much cardinality. We should probably submit a PR to reduce the collector's default config to match Tempo's.

**Reply by ogxd** [reply]
According to latest semantics version (1.24) [these attributes are now obsolete](https://github.com/open-telemetry/semantic-conventions/blob/ac0c577d20c894113eba23a344da7371cfb95a58/docs/attributes-registry/network.md#deprecated-network-attributes):
- `net.peer.name` (replaced by `server.address` on client spans and `client.address` on server spans)
- `net.sock.peer.name` (removed)
- `net.sock.peer.addr` (replaced by `network.peer.address`)
- `http.url` (replaced by `url.full`)
- `http.target` (split to `url.path` and `url.query`)

Only `db.name` and `rpc.service` are still "valid".

On the tempo side, `peer.service`, `db.name` and `db.system` are still compliant with conventions, so I think it would make sense to change the defaults of the connector to this.

**Reply by ogxd** [reply]
Regarding the virtual nodes, I have experimented with `virtual_node_peer_attributes` and I still couldn't get virtual nodes to appear. For instance if I set it to `virtual_node_peer_attributes: [peer.service]`, I can't see any additional client or server labels for this generated metric: 
`sum by (server) (rate(traces_service_graph_request_total[5m]))`

Maybe I misunderstood how this is supposed to work?

**Reply by ogxd** [reply]
Ok I think I found out, virtual nodes are under a alpha feature gate (disabled by default).

**Reply by ogxd** [reply]
I created an issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31734

**Reply by joe-elliott** [reply]
Thanks for following up on this!
